src/derivative_operators/multi_dim_bc_operators.jl

```diff
@@ -0,0 +1,161 @@
+
+    return u_temp
+end
+
+@noinline function _slice_rmul!(lower::AbstractArray, upper::AbstractArray, A::AbstractArray{B,M}, u::AbstractArray{T,N}, dim::Int, pre, post) where {T,B,N,M}
+    for J in post
+        for I in pre
+
+            tmp = A[I,J]*u[I, :, J]
+            lower[I,J], upper[I,J] = tmp.l, tmp.r
+        end
+    end
+    return (lower, upper)
+end
+
+function slice_rmul(A::AbstractArray{B,M}, u::AbstractArray{T,N}, dim::Int) where {T, B, N,M}
+    @assert N != 1
+    @assert M == N-1
+    lower = zeros(T,perpsize(u,dim))
+    upper = zeros(T,perpsize(u,dim))
+
+    _slice_rmul!(lower, upper, A, u, dim, CartesianIndices(axes(u)[1:dim-1]), CartesianIndices(axes(u)[(dim+1):end]))
+
+    return (lower, upper)
+end
+
+"""
+slicemul is the only limitation on the BCs here being used up to arbitrary dimension, an N dimensional implementation is needed.
+"""
+
+struct MultiDimDirectionalBC{T<:Number, B<:AtomicBC{T}, D, N, M} <: MultiDimensionalBC{T, N}
+    BCs::Array{B,M} #dimension M=N-1 array of BCs to extend dimension D
+end
+
+struct ComposedMultiDimBC{T, B<:AtomicBC{T}, N,M} <: MultiDimensionalBC{T, N}
+    BCs::Vector{Array{B, M}}
+end
+
+"""
+A multiple dimensional BC, supporting arbitrary BCs at each boundary point.
+To construct an arbitrary BC, pass an Array of BCs with dimension `N-1`, if `N` is the dimensionality of your domain `u`
+with a size of `size(u)[setdiff(1:N, dim)]`, where dim is the dimension orthogonal to the boundary that you want to extend.
+
+It is also possible to call
+    Q_dim = MultiDimBC(YourBC, size(u), dim)
+to use YourBC for the whole boundary orthogonal to that dimension.
+
+Further, it is possible to call
+Qx, Qy, Qz... = MultiDimBC(YourBC, size(u))
+to use YourBC for the whole boundary for all dimensions. Valid for any number of dimensions greater than 1.
+However this is only valid for Robin/General type BCs (including neummann/dirichlet) when the grid steps are equal in each dimension - including uniform grid case.
+
+In the case where you want to extend the same Robin/GeneralBC to the whole boundary with a non unifrom grid, please use
+    Qx, Qy, Qz... = RobinBC(l, r, (dx::Vector, dy::Vector, dz::Vector ...), approximation_order, size(u))
+or
+    Qx, Qy, Qz... = GeneralBC(αl, αr, (dx::Vector, dy::Vector, dz::Vector ...), approximation_order, size(u))
+
+There are also constructors for NeumannBC, DirichletBC and Dirichlet0BC. Simply replace `dx` in the call with the tuple dxyz... as above, and append `size(u)`` to the argument signature.
+The order is a required argument in this case.
+
+where dx, dy, and dz are vectors of grid steps.
+
+For Neumann0BC, please use
+    Qx, Qy, Qz... = Neumann0BC(T::Type, (dx::Vector, dy::Vector, dz::Vector ...), approximation_order, size(u))
+where T is the element type of the domain to be extended
+"""
+MultiDimBC(BC::Array{B,N}, dim::Integer) where {N, B<:AtomicBC} = MultiDimDirectionalBC{gettype(BC[1]), B, dim, N+1, N}(BC)
```

Comment:
`dim` may need to be a type parameter in order for it to infer well for dispatch

Reply:
Does something like:
```
foo(x, T) where T = bar{T}(x)
```
Work in julia? I thought that the parameters had to be inferred from argument types.

Reply:
`foo{T}(x) where T = ...